pkg/controller/s3/bucket/notificationConfig.go

```diff
@@ -54,24 +56,79 @@ func (in *NotificationConfigurationClient) Observe(ctx context.Context, bucket *
+func IsNotificationConfigurationUpToDate(cr *v1beta1.NotificationConfiguration, external *awss3.GetBucketNotificationConfigurationOutput) (ResourceStatus, error) { // nolint:gocyclo
+	// Note - aws API treats nil configuration different than empty configuration
+	// As such, we can't prealloc this due to the API
+	// If no configuration is defined but there is one in aws, we must delete it
+	if cr == nil && (len(external.QueueConfigurations) != 0 || len(external.LambdaFunctionConfigurations) != 0 || len(external.TopicConfigurations) != 0) {
+		return NeedsDeletion, nil
+	}
+	// We can't prealloc this for the API but we can to make comparison easier
+	if cr == nil {
+		cr = &v1beta1.NotificationConfiguration{
+			LambdaFunctionConfigurations: []v1beta1.LambdaFunctionConfiguration{},
+			QueueConfigurations:          []v1beta1.QueueConfiguration{},
+			TopicConfigurations:          []v1beta1.TopicConfiguration{},
+		}
+	}
+	// If any of the lengths in aws are different then there is something to delete
+	if len(cr.LambdaFunctionConfigurations) < len(external.LambdaFunctionConfigurations) || len(cr.QueueConfigurations) < len(external.QueueConfigurations) || len(cr.TopicConfigurations) < len(external.TopicConfigurations) {
+		return NeedsDeletion, nil
 	}
 
-	generated := GenerateConfiguration(config)
+	// Convert to a comparable object
+	generated := GenerateConfiguration(cr)
 
-	if cmp.Equal(external.LambdaFunctionConfigurations, generated.LambdaFunctionConfigurations, cmpopts.IgnoreTypes(document.NoSerde{})) &&
-		cmp.Equal(external.QueueConfigurations, generated.QueueConfigurations, cmpopts.IgnoreTypes(document.NoSerde{})) &&
-		cmp.Equal(external.TopicConfigurations, generated.TopicConfigurations, cmpopts.IgnoreTypes(document.NoSerde{})) {
+	// Sort everything
+	sortLambda(generated.LambdaFunctionConfigurations)
+	sortLambda(external.LambdaFunctionConfigurations)
+
+	sortQueue(generated.QueueConfigurations)
+	sortQueue(external.QueueConfigurations)
+
+	sortTopic(generated.TopicConfigurations)
+	sortTopic(external.TopicConfigurations)
+
+	if cmp.Equal(external.LambdaFunctionConfigurations, generated.LambdaFunctionConfigurations, cmpopts.IgnoreTypes(document.NoSerde{}, types.LambdaFunctionConfiguration{}.Id), cmpopts.EquateEmpty()) &&
+		cmp.Equal(external.QueueConfigurations, generated.QueueConfigurations, cmpopts.IgnoreTypes(document.NoSerde{}, types.QueueConfiguration{}.Id), cmpopts.EquateEmpty()) &&
+		cmp.Equal(external.TopicConfigurations, generated.TopicConfigurations, cmpopts.IgnoreTypes(document.NoSerde{}, types.TopicConfiguration{}.Id), cmpopts.EquateEmpty()) {
 		return Updated, nil
 	}
 
 	return NeedsUpdate, nil
 }
 
+func sortLambda(configs []types.LambdaFunctionConfiguration) {
+	sort.Slice(configs, func(i, j int) bool {
+		if a, b := configs[i].LambdaFunctionArn, configs[j].LambdaFunctionArn; a != b {
+			return aws.ToString(a) < aws.ToString(b)
+		}
+		return true
+	})
+}
```

Comment:
This is totally fine as is, but for future reference you could also pass in a couple of`cmpopts.SortSlices` options, each with a less function for a slice of the relevant type of struct.